# Heather's Anomaly Detection Project Notebook
by Heather Ortega McMillan

This is the initial notebook compiled by Heather. For full project details see the Final Notebook

In [1]:
import pandas as pd
import numpy as np

from acquire import get_joined_curriculum_data

import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings("ignore")

In [ ]:
df = get_joined_curriculum_data()

1. Which lesson appears to attract the most traffic consistently across cohorts (per program)?
    - cohort pulled from the sql table 
    - sort all the endpoints 
    
7. Which lessons are least accessed?
    - kind of related to 2

LUPE ^^^^ 

### Question I'm tackling 
2. Is there a cohort that referred to a lesson significantly more than other cohorts seemed to gloss over?
    - divided web dev and data science (4 different programs)
    - sort all the endpoints 
    - counts of cohorts accessing same endpoints
    

    

### Program ID Key 
1 - Full stack PHP

2 - Java

3 - Data Science

4 - Front End Program


In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1018810 entries, 0 to 1018809
Data columns (total 5 columns):
 #   Column     Non-Null Count    Dtype  
---  ------     --------------    -----  
 0   date       1018810 non-null  object 
 1   endpoint   1018810 non-null  object 
 2   user_id    1018810 non-null  int64  
 3   cohort_id  1018810 non-null  float64
 4   source_ip  1018810 non-null  object 
dtypes: float64(1), int64(1), object(3)
memory usage: 38.9+ MB


In [10]:
# deal with nulls by filling with 0s
#df.cohort_id.value_counts(dropna=False)

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1018810 entries, 0 to 1018809
Data columns (total 9 columns):
 #   Column      Non-Null Count    Dtype  
---  ------      --------------    -----  
 0   date        1018810 non-null  object 
 1   endpoint    1018809 non-null  object 
 2   user_id     1018810 non-null  int64  
 3   cohort_id   965313 non-null   float64
 4   source_ip   1018810 non-null  object 
 5   name        954313 non-null   object 
 6   start_date  954313 non-null   object 
 7   end_date    954313 non-null   object 
 8   program_id  954313 non-null   float64
dtypes: float64(2), int64(1), object(6)
memory usage: 77.7+ MB


In [ ]:
# Fill nulls
def fill_nulls(df):
    '''
    This function fills the nulls in the joined curriculum log data
    '''
    df = df.fillna(value = 0)
    
    return df
    
    

In [20]:
df = df.fillna(value = 0)

In [28]:
df[df.name == 0]

date                               endpoint  user_id  cohort_id  ... name start_date end_date program_id
411      2018-01-26                                      /       48        0.0  ...    0          0        0        0.0
412      2018-01-26  spring/extra-features/form-validation       48        0.0  ...    0          0        0        0.0
425      2018-01-26                                      /       48        0.0  ...    0          0        0        0.0
435      2018-01-26                                      /       48        0.0  ...    0          0        0        0.0
436      2018-01-26              mysql/relationships/joins       48        0.0  ...    0          0        0        0.0
...             ...                                    ...      ...        ...  ...  ...        ...      ...        ...
1018759  2021-07-15                  python/intro-to-numpy     1026      165.0  ...    0          0        0        0.0
1018778  2021-07-15                  python/intro-to-numpy     1051      165.0  ...    0          0        0        0.0
1018785  2021-07-15                  python/intro-to-numpy     1037      165.0  ...    0          0        0        0.0
1018786  2021-07-15                  python/intro-to-numpy     1041      165.0  ...    0          0        0        0.0
1018802  2021-07-15                  python/intro-to-numpy     1048      165.0  ...    0          0        0        0.0

[64497 rows x 9 columns]

#### Takeaways
- Newer people with no cohort id must be from Germaine and the Web Dev cohorts that just started

In [ ]:
# Is there a cohort that referred to a lesson significantly more than other cohorts seemed to 
# gloss over?


# See Lessons (endpoint) Count of each cohort access 

df['endpoint', 'cohort']



In [31]:
df[['name', 'cohort_id']].value_counts()

name        cohort_id
Staff       28.0         96954
0           0.0          53497
Ceres       33.0         40858
Jupiter     62.0         39316
Zion        29.0         38198
Fortuna     53.0         37043
Voyageurs   24.0         35707
Deimos      51.0         35381
Ganymede    57.0         34919
Apex        56.0         34172
Marco       135.0        34093
Darden      59.0         32871
Teddy       22.0         31258
Hyperion    58.0         30137
Betelgeuse  32.0         29754
Ulysses     23.0         28581
Europa      52.0         28177
Xanadu      26.0         27811
Bayes       34.0         27261
Wrangell    25.0         25680
Andromeda   31.0         25372
Kalypso     132.0        24978
Neptune     138.0        23744
Curie       55.0         22064
Florence    137.0        21435
Luna        134.0        21095
Yosemite    27.0         20864
Arches      1.0          19063
Easley      133.0        18371
Bash        61.0         18212
Oberon      139.0        17745
Lassen      14.0 

In [34]:
df[df.cohort_id == 165]

date               endpoint  user_id  cohort_id        source_ip name start_date end_date  program_id
974406   2021-06-15                      /     1026      165.0    24.242.17.211    0          0        0         0.0
974407   2021-06-15                      /     1027      165.0    69.153.21.120    0          0        0         0.0
974408   2021-06-15                      /     1028      165.0    99.66.110.207    0          0        0         0.0
974409   2021-06-15                      /     1029      165.0     67.10.172.51    0          0        0         0.0
974410   2021-06-15                      /     1030      165.0  162.238.132.211    0          0        0         0.0
...             ...                    ...      ...        ...              ...  ...        ...      ...         ...
1018759  2021-07-15  python/intro-to-numpy     1026      165.0    24.242.17.211    0          0        0         0.0
1018778  2021-07-15  python/intro-to-numpy     1051      165.0   209.182.242.80    0          0        0         0.0
1018785  2021-07-15  python/intro-to-numpy     1037      165.0  173.174.155.221    0          0        0         0.0
1018786  2021-07-15  python/intro-to-numpy     1041      165.0   70.235.249.134    0          0        0         0.0
1018802  2021-07-15  python/intro-to-numpy     1048      165.0    107.207.1.231    0          0        0         0.0

[9171 rows x 9 columns]

cohort_id 165 must be Germaine program code of 3
cohort_id 166 must be new web dev cohort (Polaris) program code of 2

+-----+---------+----------+------------+------------+---------------------+---------------------+------------+------------+
| id  | name    | slack    | start_date | end_date   | created_at          | updated_at          | deleted_at | program_id |
+-----+---------+----------+------------+------------+---------------------+---------------------+------------+------------+
| 138 | Neptune | #neptune | 2021-03-15 | 2021-09-03 | 2021-03-15 19:57:09 | 2021-03-15 19:57:09 | NULL       |          2 |
| 139 | Oberon  | #oberon  | 2021-04-12 | 2021-10-01 | 2021-04-12 18:07:21 | 2021-04-12 18:07:21 | NULL       |          2 |
| 140 | Polaris | #polaris | 2021-05-24 | 2021-11-16 | 2021-05-24 21:12:37 | 2021-05-24 21:12:37 | NULL       |          2 |
| 165 | Germain | #germain | 2021-06-14 | 2021-12-07 | 2021-06-15 13:29:36 | 2021-06-15 13:29:36 | NULL       |          3 |
| 166 | Draco   | #draco   | 2021-06-28 | 2022-01-03 | 2021-06-27 20:47:56 | 2021-06-27 20:47:56 | NULL       |          2 |
+-----+---------+----------+------------+------------+---------------------+---------------------+------------+------------+

In [41]:
# read in from clipboard copy pasted in google sheets
# Not reproduciable 
# need to put in csv for future use
cohorts2 = pd.read_clipboard(sep='|').drop(columns = ['Unnamed: 0', 'Unnamed: 10'])

In [49]:
def get_cohort_data():
    '''
    This function takes care of getting the cohort data from the SQL db
    Need env file to make work
    '''

    dbase = 'curriculum_logs'
    sqlquery = 'SELECT * FROM cohorts'

    cohorts = get_any_data(dbase, sqlquery)

    return cohorts


def get_joined_curriculum_data():
    '''
    This function joins the log data and the cohort data
    Returns one dataframe ready to be prepped. 
    '''

    df = acquire_curriculum_logs()

    df2 = get_cohort_data()

    df_merge = df.merge(df2, left_on='cohort_id', right_on= 'id', how = 'left')

    cols_to_drop = ['id', 'created_at', 'updated_at', 'deleted_at', 'slack']

    df_merge = df_merge.drop(columns = cols_to_drop)

    return df_merge

In [38]:
import acquire as a

In [39]:
cohorts1 = a.get_cohort_data()

In [40]:
cohorts1.head()

id        name        slack  start_date    end_date           created_at  \
0   1      Arches      #arches  2014-02-04  2014-04-22  2016-06-14 19:52:26   
1   2    Badlands    #badlands  2014-06-04  2014-08-22  2016-06-14 19:52:26   
2   3    Carlsbad    #carlsbad  2014-09-04  2014-11-05  2016-06-14 19:52:26   
3   4      Denali      #denali  2014-10-20  2015-01-18  2016-06-14 19:52:26   
4   5  Everglades  #everglades  2014-11-18  2015-02-24  2016-06-14 19:52:26   

            updated_at deleted_at  program_id  
0  2016-06-14 19:52:26       None           1  
1  2016-06-14 19:52:26       None           1  
2  2016-06-14 19:52:26       None           1  
3  2016-06-14 19:52:26       None           1  
4  2016-06-14 19:52:26       None           1

In [61]:
cohorts2

id     name       slack       start_date   ...   created_at             updated_at             deleted_at   program_id 
0    138   Neptune    #neptune    2021-03-15   ...   2021-03-15 19:57:09    2021-03-15 19:57:09    NULL                   2
1    139   Oberon     #oberon     2021-04-12   ...   2021-04-12 18:07:21    2021-04-12 18:07:21    NULL                   2
2    140   Polaris    #polaris    2021-05-24   ...   2021-05-24 21:12:37    2021-05-24 21:12:37    NULL                   2
3    165   Germain    #germain    2021-06-14   ...   2021-06-15 13:29:36    2021-06-15 13:29:36    NULL                   3
4    166   Draco      #draco      2021-06-28   ...   2021-06-27 20:47:56    2021-06-27 20:47:56    NULL                   2

[5 rows x 9 columns]

In [42]:
col_names = list(cohorts1)
col_names

['id',
 'name',
 'slack',
 'start_date',
 'end_date',
 'created_at',
 'updated_at',
 'deleted_at',
 'program_id']

In [43]:
# rename columns in 2nd dataframe
cohorts2.columns = col_names

In [44]:
cohorts2.info

<bound method DataFrame.info of     id       name       slack    start_date      end_date  \
0  138   Neptune    #neptune    2021-03-15    2021-09-03    
1  139   Oberon     #oberon     2021-04-12    2021-10-01    
2  140   Polaris    #polaris    2021-05-24    2021-11-16    
3  165   Germain    #germain    2021-06-14    2021-12-07    
4  166   Draco      #draco      2021-06-28    2022-01-03    

              created_at             updated_at    deleted_at  program_id  
0   2021-03-15 19:57:09    2021-03-15 19:57:09    NULL                  2  
1   2021-04-12 18:07:21    2021-04-12 18:07:21    NULL                  2  
2   2021-05-24 21:12:37    2021-05-24 21:12:37    NULL                  2  
3   2021-06-15 13:29:36    2021-06-15 13:29:36    NULL                  3  
4   2021-06-27 20:47:56    2021-06-27 20:47:56    NULL                  2  >

In [46]:
cohorts2['name']= cohorts2.name.str.strip()

In [47]:
# concat both dataframes together
full_cohorts = pd.concat([cohorts1, cohorts2], ignore_index=True, keys=list(cohorts1))

In [48]:
full_cohorts.tail(7)

id     name       slack    start_date      end_date  \
51  138  Neptune    #neptune    2021-03-15    2021-09-03   
52  139   Oberon     #oberon    2021-04-12    2021-10-01   
53  138  Neptune   #neptune    2021-03-15    2021-09-03    
54  139   Oberon   #oberon     2021-04-12    2021-10-01    
55  140  Polaris   #polaris    2021-05-24    2021-11-16    
56  165  Germain   #germain    2021-06-14    2021-12-07    
57  166    Draco   #draco      2021-06-28    2022-01-03    

               created_at             updated_at    deleted_at  program_id  
51    2021-03-15 19:57:09    2021-03-15 19:57:09          None           2  
52    2021-04-12 18:07:21    2021-04-12 18:07:21          None           2  
53   2021-03-15 19:57:09    2021-03-15 19:57:09    NULL                  2  
54   2021-04-12 18:07:21    2021-04-12 18:07:21    NULL                  2  
55   2021-05-24 21:12:37    2021-05-24 21:12:37    NULL                  2  
56   2021-06-15 13:29:36    2021-06-15 13:29:36    NULL                  3  
57   2021-06-27 20:47:56    2021-06-27 20:47:56    NULL                  2

In [49]:
# remove duplicates (oberon and neptune)
full_cohorts = full_cohorts.drop_duplicates(subset = 'id')

In [50]:
full_cohorts['name'] = full_cohorts.name.str.strip()

In [52]:
full_cohorts.to_csv('full_cohort_list.csv', index=False)

In [ ]:
def get_cohort_data():
    '''
    This function takes care of getting the cohort data from the SQL db
    Need env file to make work
    '''

    dbase = 'curriculum_logs'
    sqlquery = 'SELECT * FROM cohorts'

    cohorts = get_any_data(dbase, sqlquery)

    return cohorts


def get_full_cohort_data():
    '''
    This function takes care of getting the cohort data from the SQL db
    Need env file to make work
    '''

    cohorts = pd.read_csv('full_cohort_list.csv')

    return cohorts

In [91]:
test = pd.read_csv('full_cohort_list.csv')

id        name        slack  start_date    end_date           created_at           updated_at deleted_at  program_id
0   1      Arches      #arches  2014-02-04  2014-04-22  2016-06-14 19:52:26  2016-06-14 19:52:26        NaN           1
1   2    Badlands    #badlands  2014-06-04  2014-08-22  2016-06-14 19:52:26  2016-06-14 19:52:26        NaN           1
2   3    Carlsbad    #carlsbad  2014-09-04  2014-11-05  2016-06-14 19:52:26  2016-06-14 19:52:26        NaN           1
3   4      Denali      #denali  2014-10-20  2015-01-18  2016-06-14 19:52:26  2016-06-14 19:52:26        NaN           1
4   5  Everglades  #everglades  2014-11-18  2015-02-24  2016-06-14 19:52:26  2016-06-14 19:52:26        NaN           1

In [3]:
df = get_joined_curriculum_data()

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1018810 entries, 0 to 1018809
Data columns (total 5 columns):
 #   Column     Non-Null Count    Dtype  
---  ------     --------------    -----  
 0   date       1018810 non-null  object 
 1   endpoint   1018810 non-null  object 
 2   user_id    1018810 non-null  int64  
 3   cohort_id  1018810 non-null  float64
 4   source_ip  1018810 non-null  object 
dtypes: float64(1), int64(1), object(3)
memory usage: 38.9+ MB


In [5]:
df[['name', 'cohort_id']].value_counts()

KeyError: "['name'] not in index"

In [10]:
df = df.fillna(value = 0)

In [11]:
df[['name', 'cohort_id']].value_counts()

name        cohort_id
Staff       28.0         96954
0           0.0          53497
Ceres       33.0         40858
Jupiter     62.0         39316
Zion        29.0         38198
Fortuna     53.0         37043
Voyageurs   24.0         35707
Deimos      51.0         35381
Ganymede    57.0         34919
Apex        56.0         34172
Marco       135.0        34093
Darden      59.0         32871
Teddy       22.0         31258
Hyperion    58.0         30137
Betelgeuse  32.0         29754
Ulysses     23.0         28581
Europa      52.0         28177
Xanadu      26.0         27811
Bayes       34.0         27261
Wrangell    25.0         25680
Andromeda   31.0         25372
Kalypso     132.0        24978
Neptune     138.0        23744
Curie       55.0         22064
Florence    137.0        21435
Luna        134.0        21095
Yosemite    27.0         20864
Arches      1.0          19063
Easley      133.0        18371
Bash        61.0         18212
Oberon      139.0        17745
Lassen      14.0 

In [12]:
df[df.name == 0]

date                               endpoint  user_id  \
411      2018-01-26                                      /       48   
412      2018-01-26  spring/extra-features/form-validation       48   
425      2018-01-26                                      /       48   
435      2018-01-26                                      /       48   
436      2018-01-26              mysql/relationships/joins       48   
...             ...                                    ...      ...   
1013475  2021-07-12      java-i/syntax-types-and-variables      810   
1014540  2021-07-13                                      /      727   
1015184  2021-07-13                                      /      728   
1017791  2021-07-14                                      /      810   
1017792  2021-07-14                                      /      810   

         cohort_id       source_ip name start_date end_date  program_id  
411            0.0    97.105.19.61    0          0        0         0.0  
412            0.0    97.105.19.61    0          0        0         0.0  
425            0.0    97.105.19.61    0          0        0         0.0  
435            0.0    97.105.19.61    0          0        0         0.0  
436            0.0    97.105.19.61    0          0        0         0.0  
...            ...             ...  ...        ...      ...         ...  
1013475        0.0   162.92.128.83    0          0        0         0.0  
1014540        0.0   172.58.96.119    0          0        0         0.0  
1015184        0.0  70.130.121.144    0          0        0         0.0  
1017791        0.0     75.28.22.72    0          0        0         0.0  
1017792        0.0     75.28.22.72    0          0        0         0.0  

[53497 rows x 9 columns]

In [33]:
df['name'] = df.name.str.strip()

In [34]:
df.name.value_counts()

Staff         96954
Ceres         40858
Jupiter       39316
Zion          38198
Fortuna       37043
Voyageurs     35707
Deimos        35381
Ganymede      34919
Apex          34172
Marco         34093
Darden        32871
Teddy         31258
Hyperion      30137
Betelgeuse    29754
Ulysses       28581
Europa        28177
Xanadu        27811
Bayes         27261
Wrangell      25680
Andromeda     25372
Kalypso       24978
Neptune       23744
Curie         22064
Florence      21435
Luna          21095
Yosemite      20864
Arches        19063
Easley        18371
Bash          18212
Oberon        17745
Lassen         9596
Germain        9171
Sequoia        7444
Olympic        4960
Kings          2848
Pinnacles      2257
Draco          1829
Hampton        1712
Quincy         1371
Glacier         833
Niagara         755
Mammoth         691
Joshua          302
Ike             253
Badlands         95
Franklin         72
Apollo            5
Denali            4
Everglades        1
Name: name, dtype: i

In [35]:
df[df['name'] == 'Germain']

date               endpoint  user_id  cohort_id  \
974406   2021-06-15                      /     1026      165.0   
974407   2021-06-15                      /     1027      165.0   
974408   2021-06-15                      /     1028      165.0   
974409   2021-06-15                      /     1029      165.0   
974410   2021-06-15                      /     1030      165.0   
...             ...                    ...      ...        ...   
1018759  2021-07-15  python/intro-to-numpy     1026      165.0   
1018778  2021-07-15  python/intro-to-numpy     1051      165.0   
1018785  2021-07-15  python/intro-to-numpy     1037      165.0   
1018786  2021-07-15  python/intro-to-numpy     1041      165.0   
1018802  2021-07-15  python/intro-to-numpy     1048      165.0   

               source_ip     name    start_date      end_date  program_id  
974406     24.242.17.211  Germain   2021-06-14    2021-12-07          3.0  
974407     69.153.21.120  Germain   2021-06-14    2021-12-07          3.0  
974408     99.66.110.207  Germain   2021-06-14    2021-12-07          3.0  
974409      67.10.172.51  Germain   2021-06-14    2021-12-07          3.0  
974410   162.238.132.211  Germain   2021-06-14    2021-12-07          3.0  
...                  ...      ...           ...           ...         ...  
1018759    24.242.17.211  Germain   2021-06-14    2021-12-07          3.0  
1018778   209.182.242.80  Germain   2021-06-14    2021-12-07          3.0  
1018785  173.174.155.221  Germain   2021-06-14    2021-12-07          3.0  
1018786   70.235.249.134  Germain   2021-06-14    2021-12-07          3.0  
1018802    107.207.1.231  Germain   2021-06-14    2021-12-07          3.0  

[9171 rows x 9 columns]

In [37]:
df.to_csv('full_cohort_list.csv', index = False)